library/file.lisp

```diff
@@ -0,0 +1,663 @@
+  (define (remove-directory-unsafe path)
+    "Deletes a target directory recursively. Equivalent to `rm -r`. Errors if the path is not a directory."
+    (if-directory-path path
+                       (fn (p)
+                         (lisp (Result FileError Unit) (p)
+                           (handle-file-function (uiop:delete-directory-tree p))))))
+  
+  (declare system-relative-pathname ((Into :a String) => :a -> String -> (Result FileError Pathname)))
+  (define (system-relative-pathname system-name name)
+    "Generates a system-relative-pathname for a given filename or path."
+    (lisp (Result FileError Pathname) (system-name name)
+      (cl:handler-case (Ok (asdf:system-relative-pathname system-name name))
+        (cl:error (c) (Err (LispError c)))))))
+
+;;;
+;;; Basic File Operations
+;;;
+
+(coalton-toplevel
+
+  (declare copy ((Into :a Pathname) (Into :b Pathname) => :a -> :b -> (Result FileError :c )))
+  (define (copy input output)
+    "Copies a file to a nonexistant file. Will error if the output path already exists or if the input path isn't a file."
+    (do
+     (let in = (the Pathname (into input)))
+     (let out = (the Pathname (into output)))
+      (output-exists <- (exists? out))
+      (input-is-file <- (valid-file-pathname? in))
+      (if output-exists
+          (Err (PathError "Invalid output for copying, path already exists:" out))
+          (if (not input-is-file)
+              (Err (PathError "Invalid input for copying, path is not a file:" in))
+              (lisp (Result FileError :c) (in out)
+                (handle-file-function (uiop:copy-file in out)))))))
+
+  (declare delete-file ((Into :a Pathname) => :a -> (Result FileError Unit)))
+  (define (delete-file path)
+    "Deletes a given file if the file exists."
+    (do
+     (let p = (the Pathname (into path)))
+     (lisp (Result FileError Unit) (p)
+       (handle-file-function (uiop:delete-file-if-exists p))))))
+
+;;;
+;;; FStreams, FileStreams, and options
+;;;
+
+(coalton-toplevel
+
+  (repr :native cl:file-stream)
+  (define-type (FileStream :a)
+    "Represents a file stream, using `cl:file-stream`.")
+
+  (repr :enum)
+  (define-type IfExists
+    "Possible options for opening a stream when the file exists."
+    EError
+    NewVersion
+    Rename
+    RenameAndDelete
+    Overwrite
+    Append
+    Supersede)
+
+  (repr :enum)
+  (define-type IfDoesNotExist
+    "Possible options for opening a stream when the file does not exist."
+    DNEError
+    Create)
+
+  (define-type StreamOptions
+    "A type for providing parameters for opening streams. StreamOptions take strings for pathnames, but they will error if they are not proper and appropriate pathnames."
+    (Input Pathname)
+    (Output Pathname IfExists IfDoesNotExist)
+    (TwoWay Pathname IfExists IfDoesNotExist))
+
+  ;;
+  ;; Opening Streams
+  ;;
+  (declare %open-input (Pathname -> types:lisptype -> (Result FileError (FileStream :a))))
```

Comment:
I think you should use a type class to be generic on the type of `FileStream`.

Reply:
Does the `Streamable` class satisfy this? These  %open... functions are just helper functions for the type class.